File: go/internal/realtime/protocol.go
package realtime

import (
	"encoding/json"
	"errors"
	"fmt"
)

const ProtocolVersion = 1

type Envelope struct {
	Version int             `json:"v"`
	Type    string          `json:"type"`
	ID      string          `json:"id,omitempty"`
	Payload json.RawMessage `json:"payload,omitempty"`
}

type HelloPayload struct {
	ConnectionID    string `json:"connectionId"`
	ProtocolVersion int    `json:"protocolVersion"`
	HeartbeatMillis int64  `json:"heartbeatMillis"`
	ResumeSupported bool   `json:"resumeSupported"`
}

type HeartbeatPayload struct {
	Sequence uint64 `json:"sequence"`
}

type ResumePayload struct {
	Token        string `json:"token"`
	LastSequence uint64 `json:"lastSequence"`
}

type ErrorPayload struct {
	Code          string `json:"code"`
	Message       string `json:"message"`
	CorrelationID string `json:"correlationId,omitempty"`
}

func Decode(data []byte) (Envelope, error) {
	var envelope Envelope
	if err := json.Unmarshal(data, &envelope); err != nil {
		return Envelope{}, errors.New("invalid JSON")
	}
	if envelope.Version != ProtocolVersion {
		return Envelope{}, fmt.Errorf("unsupported protocol version %d", envelope.Version)
	}
	if envelope.Type == "" {
		return Envelope{}, errors.New("message type is required")
	}
	return envelope, nil
}

func NewEnvelope(messageType, id string, payload any) (Envelope, error) {
	data, err := json.Marshal(payload)
	if err != nil {
		return Envelope{}, err
	}
	return Envelope{Version: ProtocolVersion, Type: messageType, ID: id, Payload: data}, nil
}

File: go/web/src/lib/realtime/protocol.ts
export const protocolVersion = 1;

export type MessageType =
  | 'connection.hello'
  | 'connection.resume'
  | 'connection.resumed'
  | 'connection.resume_unavailable'
  | 'heartbeat.ping'
  | 'heartbeat.pong'
  | 'error';

export interface Envelope<T = unknown> {
  v: number;
  type: MessageType;
  id?: string;
  payload?: T;
}

export interface HelloPayload {
  connectionId: string;
  protocolVersion: number;
  heartbeatMillis: number;
  resumeSupported: boolean;
}

export interface HeartbeatPayload {
  sequence: number;
}

const messageTypes = new Set<MessageType>([
  'connection.hello',
  'connection.resume',
  'connection.resumed',
  'connection.resume_unavailable',
  'heartbeat.ping',
  'heartbeat.pong',
  'error',
]);

export function parseEnvelope(value: string): Envelope {
  const parsed: unknown = JSON.parse(value);
  if (!isRecord(parsed) || parsed.v !== protocolVersion || typeof parsed.type !== 'string') {
    throw new Error('Invalid realtime envelope');
  }
  if (!messageTypes.has(parsed.type as MessageType)) {
    throw new Error('Unsupported realtime message');
  }
  if (parsed.id !== undefined && typeof parsed.id !== 'string') {
    throw new Error('Invalid realtime message id');
  }
  return parsed as unknown as Envelope;
}

export function isHelloPayload(payload: unknown): payload is HelloPayload {
  return (
    isRecord(payload) &&
    typeof payload.connectionId === 'string' &&
    payload.protocolVersion === protocolVersion &&
    typeof payload.heartbeatMillis === 'number' &&
    payload.heartbeatMillis > 0 &&
    typeof payload.resumeSupported === 'boolean'
  );
}

export function isHeartbeatPayload(payload: unknown): payload is HeartbeatPayload {
  return (
    isRecord(payload) &&
    typeof payload.sequence === 'number' &&
    Number.isSafeInteger(payload.sequence) &&
    payload.sequence >= 0
  );
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}
